README.md

```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+
+- Remove the previous effect and move its content to a new function `fetchFeed`.
+  You'll need to modify the URL to dynamically pass the `username`, this can be
+  done by leveraging string concatenation operator `++` (see the [syntax
+  cheatsheet](https://reasonml.github.io/docs/en/syntax-cheatsheet)).
+- Create a new effect that has `username` as a dependency (hint: use
+  `React.useEffect1`)
+- Inside the effect, make sure to:
+    - clear the existing timeout `debounceTimeout` if it has a value set (its
+      value is `Some(_)`) using
+      [Js.Global.clearTimeout](https://melange.re/v4.0.0/api/re/melange/Js/Global/#val-clearTimeout)
+    - set a new timeout using
+      [Js.Global.setTimeout](https://melange.re/v4.0.0/api/re/melange/Js/Global/#val-setTimeout).
+      Let's set it to 500ms.
+    - store the new timeout id using `setDebounceTimeout` function from the
+      `useState` hook
+
+### Step 8 completion check
+
+After this step is completed, we should be able to type a valid username like
+`xavierleroy` and see their most recent activity on GitHub appearing in the
+feed.
+
+Don't worry about data validation or error handling for now, we will fix that in
+the next step.
+
+## Step 9: Validating data with abstract modules, handling edge cases and errors
+
+### 9.1: Validate usernames
+
+Right now, we can enter any kind of string in our input, but GitHub only allows
+usernames with specific rules:
+- Max length should be 39 characters
+- Can only contain alphanumeric characters and dashes `-`
+- Can't start or finish with dash `-`
+
+Let's add some logic to validate the username. As the `App.re` module is getting
+large already, let's create a new module `Username.re`. In this module, we will
+define the following:
+
+- `type t = string`
+- A function `make` that takes a `string` and returns a `result` value:
+```reason
+let make = (username: string) => {
+  let re = [%re "/^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,37}[a-zA-Z0-9])?$/"];
+  Js.Re.test(~str=username, re) ? Ok(username) : Error();
+};
+```
+- A `toString` function that converts `Username.t` back to a string.
+
+To make sure we can't use any string as validated username incorrectly, we will
+make the type `t`
+[abstract](https://courses.cs.cornell.edu/cs3110/2021sp/textbook/modules/abstract_types.html),
+In order to do so, we will add an [interface
+file](https://reasonml.github.io/docs/en/module#interface-files). These files
+define how the module types will be visible from the outside.
+
+So let's create a new file `Username.rei` with this content:
+
+```reasonml
+type t;
+
+let make: string => result(t, unit);
+
+let toString: t => string;
+```
+Note how we are leaving `t` without a definition, this makes the type abstract.
+This way, every time we need to use a GitHub username in our app, we can use
+`Username.t` as a type, and this will guarantee that the string has gone through
+the validation before being used in the function.
```

Comment:
I would explain verbally the difference between TypeScript and OCaml regarding nominal and structural. Those details are hard to grasp without an example, but this is a great example to showcase.

Reply:
also, can explain that running a regex might be silly and you might want to run this "validation" async by hitting their API

Reply:
> I would explain verbally the difference between TypeScript and OCaml regarding nominal and structural. Those details are hard to grasp without an example, but this is a great example to showcase.

I don't see how nominal types enable abstract types 🤔 it's rather the separation between interface and implementation no? see e.g. https://github.com/microsoft/TypeScript/issues/15408#issuecomment-298162519. Maybe i don't understand fully, can you elaborate?

> can explain that running a regex might be silly and you might want to run this "validation" async by hitting their API

I can't find any GitHub API endpoints to validate a username, do you know how?